<a href="https://colab.research.google.com/github/szabeebzaliz7-a11y/Datascience-And-Gen-AI/blob/main/Assignment_06_Linear_Regression.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

Regression Assignment Objective:
The objective of this assignment is to evaluate your understanding of regression techniques in
supervised learning by applying them to a real-world dataset.

Dataset:
Use the California Housing dataset available in the sklearn library. This dataset contains
information about various features of houses in California and their respective median prices.

1. Loading and Preprocessing :
● Load the California Housing dataset using the fetch_california_housing function
from sklearn.
● Convert the dataset into a pandas DataFrame for easier handling. Handle
missing values (if any) and perform necessary feature scaling (e.g.,
standardization).
● Explain the preprocessing steps you performed and justify why they are
necessary for this dataset.

In [21]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
from sklearn.datasets import fetch_california_housing

In [22]:
housing = fetch_california_housing()

df = pd.DataFrame(housing.data, columns=housing.feature_names)
df['MedHouseVal'] = housing.target

In [23]:
df

,MedInc,HouseAge,AveRooms,AveBedrms,Population,AveOccup,Latitude,Longitude,MedHouseVal
0,8.3252,41.0,6.984127,1.023810,322.0,2.555556,37.88,-122.23,4.526
1,8.3014,21.0,6.238137,0.971880,2401.0,2.109842,37.86,-122.22,3.585
2,7.2574,52.0,8.288136,1.073446,496.0,2.802260,37.85,-122.24,3.521
3,5.6431,52.0,5.817352,1.073059,558.0,2.547945,37.85,-122.25,3.413
4,3.8462,52.0,6.281853,1.081081,565.0,2.181467,37.85,-122.25,3.422
...,...,...,...,...,...,...,...,...,...
20635,1.5603,25.0,5.045455,1.133333,845.0,2.560606,39.48,-121.09,0.781
20636,2.5568,18.0,6.114035,1.315789,356.0,3.122807,39.49,-121.21,0.771
20637,1.7000,17.0,5.205543,1.120092,1007.0,2.325635,39.43,-121.22,0.923
20638,1.8672,18.0,5.329513,1.171920,741.0,2.123209,39.43,-121.32,0.847


In [24]:
print(df.info())
print(df.isnull().sum())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 20640 entries, 0 to 20639
Data columns (total 9 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   MedInc       20640 non-null  float64
 1   HouseAge     20640 non-null  float64
 2   AveRooms     20640 non-null  float64
 3   AveBedrms    20640 non-null  float64
 4   Population   20640 non-null  float64
 5   AveOccup     20640 non-null  float64
 6   Latitude     20640 non-null  float64
 7   Longitude    20640 non-null  float64
 8   MedHouseVal  20640 non-null  float64
dtypes: float64(9)
memory usage: 1.4 MB
None
MedInc         0
HouseAge       0
AveRooms       0
AveBedrms      0
Population     0
AveOccup       0
Latitude       0
Longitude      0
MedHouseVal    0
dtype: int64


In [25]:
from sklearn.preprocessing import StandardScaler

In [26]:
standard_scaler = StandardScaler()
standard_scaled_data=standard_scaler.fit_transform(df)
standard_scaled_data

array([[ 2.34476576,  0.98214266,  0.62855945, ...,  1.05254828,
        -1.32783522,  2.12963148],
       [ 2.33223796, -0.60701891,  0.32704136, ...,  1.04318455,
        -1.32284391,  1.31415614],
       [ 1.7826994 ,  1.85618152,  1.15562047, ...,  1.03850269,
        -1.33282653,  1.25869341],
       ...,
       [-1.14259331, -0.92485123, -0.09031802, ...,  1.77823747,
        -0.8237132 , -0.99274649],
       [-1.05458292, -0.84539315, -0.04021111, ...,  1.77823747,
        -0.87362627, -1.05860847],
       [-0.78012947, -1.00430931, -0.07044252, ...,  1.75014627,
        -0.83369581, -1.01787803]])

2. Regression Algorithm Implementation:
● Implement the following regression algorithms:
○ Linear Regression
○ Decision Tree Regressor
○ Random Forest Regressor
○ Gradient Boosting Regressor
○ Support Vector Regressor (SVR)
● For each algorithm: Provide a brief explanation of how it works. Explain why it
might be suitable for this dataset.

In [27]:
from sklearn.linear_model import LinearRegression
from sklearn.tree import DecisionTreeRegressor
from sklearn.ensemble import RandomForestRegressor, GradientBoostingRegressor
from sklearn.svm import SVR

In [28]:
from sklearn.model_selection import train_test_split


In [29]:
lnr=LinearRegression()
x=df.drop('MedHouseVal',axis=1)
y=df['MedHouseVal']
X_train, X_test, y_train, y_test=train_test_split(x, y, random_state=42)
lnr.fit(X_train,y_train)
print(lnr.score(X_train,y_train))


0.609873031052925


In [30]:
rf = RandomForestRegressor()
rf.fit(X_train, y_train)
print(rf.score(X_train,y_train))


0.9730387113519495


In [31]:
dt=DecisionTreeRegressor()
dt.fit(X_train, y_train)
print(dt.score(X_train,y_train))


1.0


In [32]:
gdb=GradientBoostingRegressor()
gdb.fit(X_train, y_train)
print(gdb.score(X_train,y_train))


0.8061675157085277


In [14]:
svcr=SVR()
svcr.fit(X_train, y_train)
print(svcr.score(X_train,y_train))


-0.02670539815031714


3. Model Evaluation and Comparison:
● Evaluate the performance of each algorithm using the following metrics:
○ Mean Squared Error (MSE)
○ Mean Absolute Error (MAE)
○ R-squared Score (R2)
● Compare the results of all models and identify: The best-performing algorithm
with justification and the worst-performing algorithm with reasoning.

In [15]:
from sklearn.metrics import mean_squared_error, mean_absolute_error , r2_score

In [16]:
predictions = lnr.predict(X_test)
print("Mean Squared Error:", mean_squared_error(y_test, predictions))
print("Mean Absolute Error:", mean_absolute_error(y_test, predictions))
print("R-squared Score:", r2_score(y_test, predictions))

Mean Squared Error: 0.5411287478470685
Mean Absolute Error: 0.5296964012919464
R-squared Score: 0.5910509795491354


In [17]:
predictions = rf.predict(X_test)
print("Mean Squared Error:", mean_squared_error(y_test, predictions))
print("Mean Absolute Error:", mean_absolute_error(y_test, predictions))
print("R-squared Score:", r2_score(y_test, predictions))

Mean Squared Error: 0.2532463405134531
Mean Absolute Error: 0.3270628626356591
R-squared Score: 0.8086133045087972


In [18]:
predictions = dt.predict(X_test)
print("Mean Squared Error:", mean_squared_error(y_test, predictions))
print("Mean Absolute Error:", mean_absolute_error(y_test, predictions))
print("R-squared Score:", r2_score(y_test, predictions))

Mean Squared Error: 0.5304181197834495
Mean Absolute Error: 0.46734132945736434
R-squared Score: 0.5991453579617723


In [19]:
predictions = gdb.predict(X_test)
print("Mean Squared Error:", mean_squared_error(y_test, predictions))
print("Mean Absolute Error:", mean_absolute_error(y_test, predictions))
print("R-squared Score:", r2_score(y_test, predictions))

Mean Squared Error: 0.2893789978245941
Mean Absolute Error: 0.37110935460211214
R-squared Score: 0.7813066517529286


In [20]:
predictions = svcr.predict(X_test)
print("Mean Squared Error:", mean_squared_error(y_test, predictions))
print("Mean Absolute Error:", mean_absolute_error(y_test, predictions))
print("R-squared Score:", r2_score(y_test, predictions))

Mean Squared Error: 1.3525489127697092
Mean Absolute Error: 0.8674205743031635
R-squared Score: -0.02216626854460868


### Model Comparison:

**Linear Regression:**
*   Mean Squared Error: 0.541
*   Mean Absolute Error: 0.530
*   R-squared Score: 0.591

**Decision Tree Regressor:**
*   Mean Squared Error: 0.517
*   Mean Absolute Error: 0.463
*   R-squared Score: 0.609

**Random Forest Regressor:**
*   Mean Squared Error: 0.254
*   Mean Absolute Error: 0.329
*   R-squared Score: 0.808

**Gradient Boosting Regressor:**
*   Mean Squared Error: 0.297
*   Mean Absolute Error: 0.371
*   R-squared Score: 0.781

**Support Vector Regressor (SVR):**
*   Mean Squared Error: 1.353
*   Mean Absolute Error: 0.867
*   R-squared Score: -0.022

### Analysis:

**Best Performing Algorithm: Random Forest Regressor**

The Random Forest Regressor stands out as the best-performing algorithm. It has the **lowest Mean Squared Error (0.254)** and **Mean Absolute Error (0.329)**, indicating that its predictions are closest to the actual values on average. Crucially, it also boasts the **highest R-squared score (0.808)**. An R-squared value of 0.808 means that approximately 80.8% of the variance in the target variable (Median House Value) can be explained by the independent variables in the model. This high R-squared, coupled with low errors, suggests a robust model that fits the data very well and generalizes effectively to unseen data.

**Worst Performing Algorithm: Support Vector Regressor (SVR)**

The Support Vector Regressor (SVR) is the worst-performing algorithm in this comparison. It exhibits the **highest Mean Squared Error (1.353)** and **Mean Absolute Error (0.867)**, meaning its predictions have the largest deviations from the actual values. More notably, its **R-squared score is -0.022**. A negative R-squared score indicates that the model performs worse than a simple horizontal line (the mean of the target variable). This suggests that the SVR model, in its current configuration, is not capturing any meaningful patterns in the data and is effectively a very poor fit for this dataset.